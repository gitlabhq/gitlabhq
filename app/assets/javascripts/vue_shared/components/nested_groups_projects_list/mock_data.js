/* eslint-disable @gitlab/require-i18n-strings */
import { uniqueId } from 'lodash';
import { ACTION_EDIT, ACTION_DELETE } from '~/vue_shared/components/list_actions/constants';
import { slugify } from '~/lib/utils/text_utility';
import { LIST_ITEM_TYPE_PROJECT, LIST_ITEM_TYPE_GROUP } from './constants';

const makeGroup = ({ name, fullName, childrenToLoad = [] }) => {
  const fullPath = slugify(fullName);
  const id = parseInt(uniqueId(), 10);

  return {
    type: LIST_ITEM_TYPE_GROUP,
    markedForDeletionOn: null,
    permanentDeletionDate: '2025-02-26',
    fullPath,
    descriptionHtml:
      '<p data-sourcepos="1:1-1:64" dir="auto">Sapiente excepturi est eos corrupti possimus praesentium quidem.</p>',
    avatarUrl: null,
    descendantGroupsCount: 0,
    projectsCount: 1,
    groupMembersCount: 5,
    visibility: 'public',
    createdAt: '2024-09-05T11:04:39Z',
    updatedAt: '2024-10-03T18:09:02Z',
    isLinkedToSubscription: true,
    id,
    avatarLabel: name,
    fullName,
    webUrl: `https://gdk.test:3443/${fullPath}`,
    accessLevel: { integerValue: 50 },
    editPath: `/${fullPath}/edit`,
    availableActions: [ACTION_EDIT, ACTION_DELETE],
    children: [],
    childrenToLoad: childrenToLoad.map((child) => ({ ...child, parent: { id } })),
    hasChildren: childrenToLoad.length > 0,
    childrenCount: childrenToLoad.length,
    childrenLoading: false,
  };
};

const makeProject = ({ name, nameWithNamespace }) => {
  const fullPath = slugify(nameWithNamespace);

  return {
    type: LIST_ITEM_TYPE_PROJECT,
    markedForDeletionOn: null,
    permanentDeletionDate: '2025-02-26',
    fullPath,
    archived: false,
    webUrl: `https://gdk.test:3443/${fullPath}`,
    topics: [],
    forksCount: 1,
    avatarUrl: null,
    starCount: 1,
    visibility: 'public',
    openMergeRequestsCount: 4,
    openIssuesCount: 47,
    descriptionHtml:
      '<p data-sourcepos="1:1-1:503" dir="auto">Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam.</p>',
    createdAt: '2024-09-05T11:04:37Z',
    updatedAt: '2024-12-12T18:33:54Z',
    lastActivityAt: '2024-12-12T18:33:54Z',
    userPermissions: {
      removeProject: true,
      viewEditPage: true,
    },
    isCatalogResource: false,
    exploreCatalogPath: null,
    pipeline: {
      detailedStatus: {
        id: 'failed-566-566',
        icon: 'status_failed',
        text: 'Failed',
        detailsPath: `/${fullPath}/-/pipelines/566`,
      },
    },
    id: parseInt(uniqueId(), 10),
    name,
    nameWithNamespace,
    avatarLabel: name,
    mergeRequestsAccessLevel: 'ENABLED',
    issuesAccessLevel: 'ENABLED',
    forkingAccessLevel: 'ENABLED',
    isForked: false,
    accessLevel: { integerValue: 50 },
    availableActions: [ACTION_EDIT, ACTION_DELETE],
    editPath: `/${fullPath}/edit`,
  };
};

export const projectA = makeProject({
  name: 'Project A',
  nameWithNamespace: 'Top-level group A / Subgroup / Nested subgroup / Project A',
});

export const projectB = makeProject({
  name: 'Project B',
  nameWithNamespace: 'Top-level group A / Project B',
});

export const nestedSubgroup = makeGroup({
  name: 'Nested subgroup',
  fullName: 'Top-level group A / Subgroup / Nested subgroup',
  childrenToLoad: [projectA],
});

export const subgroup = makeGroup({
  name: 'Subgroup',
  fullName: 'Top-level group A / Subgroup',
  childrenToLoad: [nestedSubgroup],
});

export const topLevelGroupA = makeGroup({
  name: 'Top-level group A',
  fullName: 'Top-level group A',
  childrenToLoad: [subgroup, projectB],
});

export const topLevelGroupB = makeGroup({
  name: 'Top-level group B',
  fullName: 'Top-level group B',
});

export const items = [topLevelGroupA, topLevelGroupB];
