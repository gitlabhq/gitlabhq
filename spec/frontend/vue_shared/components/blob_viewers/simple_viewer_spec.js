import Vue, { nextTick } from 'vue';
import VueApollo from 'vue-apollo';
import { shallowMount } from '@vue/test-utils';
import createMockApollo from 'helpers/mock_apollo_helper';
import {
  HIGHLIGHT_CLASS_NAME,
  MAX_BLAME_LINES,
} from '~/vue_shared/components/blob_viewers/constants';
import SimpleViewer from '~/vue_shared/components/blob_viewers/simple_viewer.vue';
import waitForPromises from 'helpers/wait_for_promises';

import blameDataQuery from '~/vue_shared/components/source_viewer/queries/blame_data.query.graphql';
import Blame from '~/vue_shared/components/source_viewer/components/blame_info.vue';

import { BLAME_DATA_QUERY_RESPONSE_MOCK } from './mock_data';

Vue.use(VueApollo);

describe('Blob Simple Viewer component', () => {
  let wrapper;
  let fakeApollo;
  const contentMock = `<span id="LC1">First</span>\n<span id="LC2">Second</span>\n<span id="LC3">Third</span>`;
  const blobHash = 'foo-bar';

  const blameDataQueryHandlerSuccess = jest.fn().mockResolvedValue(BLAME_DATA_QUERY_RESPONSE_MOCK);

  function createComponent({
    content = contentMock,
    showBlame = false,
    blamePath = '/blame/foo-bar',
    isBlameLinkHidden = false,
    isRawContent = false,
    propsData = {},
  } = {}) {
    fakeApollo = createMockApollo([[blameDataQuery, blameDataQueryHandlerSuccess]]);

    wrapper = shallowMount(SimpleViewer, {
      apolloProvider: fakeApollo,
      provide: {
        blobHash,
      },
      propsData: {
        content,
        type: 'text',
        fileName: 'test.js',
        isRawContent,
        isBlameLinkHidden,
        blamePath,
        blobPath: 'podfile',
        projectPath: 'test',
        currentRef: 'test',
        lineNumbers: 3,
        showBlame,
        ...propsData,
      },
    });
  }
  const findBlameComponents = () => wrapper.findAllComponents(Blame);
  const findBlameLinks = () => wrapper.findAll('.file-line-blame');

  it('does not fail if content is empty', () => {
    const spy = jest.spyOn(window.console, 'error');
    createComponent({ content: '' });
    expect(spy).not.toHaveBeenCalled();
  });

  describe('rendering', () => {
    beforeEach(() => {
      createComponent();
    });

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('renders exactly three lines', () => {
      expect(wrapper.findAll('.file-line-num')).toHaveLength(3);
    });

    it('renders a blame link for each line with correct href attribute', () => {
      expect(findBlameLinks()).toHaveLength(3);
      expect(findBlameLinks().at(0).attributes('href')).toBe('/blame/foo-bar#L1');
    });

    it('does not render blame link when `isBlameLinkHidden` prop set to true', () => {
      createComponent({ isBlameLinkHidden: true });
      expect(findBlameLinks()).toHaveLength(0);
    });

    it('renders the content without transformations', () => {
      expect(wrapper.html()).toContain(contentMock);
    });

    it('does not render a Blame component when `showBlame: false`', async () => {
      await waitForPromises();
      expect(findBlameComponents()).toHaveLength(0);
    });

    describe('Blame component', () => {
      beforeEach(() => {
        createComponent({ propsData: { showBlame: true } });
      });
      it('renders a Blame component with correct props', async () => {
        await waitForPromises();
        const blameInfo =
          BLAME_DATA_QUERY_RESPONSE_MOCK.data.project.repository.blobs.nodes[0].blame.groups;

        expect(findBlameComponents().at(0).exists()).toBe(true);
        expect(findBlameComponents().at(0).props()).toMatchObject({ blameInfo });
      });

      it('calls the blame data query', async () => {
        await waitForPromises();
        expect(blameDataQueryHandlerSuccess).toHaveBeenCalledTimes(1);
        expect(blameDataQueryHandlerSuccess).toHaveBeenCalledWith(
          expect.objectContaining({
            filePath: 'podfile',
            fromLine: 1,
            fullPath: 'test',
            ref: 'test',
            toLine: MAX_BLAME_LINES,
          }),
        );
      });

      it('preloads blame data', async () => {
        jest.clearAllMocks();
        createComponent({ propsData: { showBlame: false, shouldPreloadBlame: true } });
        await waitForPromises();

        expect(blameDataQueryHandlerSuccess).toHaveBeenCalledWith(
          expect.objectContaining({
            filePath: 'podfile',
            fromLine: 1,
            fullPath: 'test',
            ref: 'test',
            toLine: MAX_BLAME_LINES,
          }),
        );
      });

      it('if there is more than 100 lines it calls query as many time as needed', async () => {
        createComponent({ propsData: { showBlame: true, lineNumbers: 201 } });
        await waitForPromises();
        await waitForPromises();
        expect(blameDataQueryHandlerSuccess).toHaveBeenCalledTimes(4);
      });
    });
  });

  describe('functionality', () => {
    const scrollIntoViewMock = jest.fn();
    HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

    beforeEach(() => {
      window.location.hash = '#LC2';
      createComponent();
    });

    afterEach(() => {
      window.location.hash = '';
    });

    it('scrolls to requested line when rendered', () => {
      const linetoBeHighlighted = wrapper.find('#LC2');
      expect(scrollIntoViewMock).toHaveBeenCalled();
      expect(wrapper.vm.highlightedLine).toBe(linetoBeHighlighted.element);
      expect(linetoBeHighlighted.classes()).toContain(HIGHLIGHT_CLASS_NAME);
    });

    it('switches highlighting when another line is selected', async () => {
      const currentlyHighlighted = wrapper.find('#LC2');
      const hash = '#LC3';
      const linetoBeHighlighted = wrapper.find(hash);

      expect(wrapper.vm.highlightedLine).toBe(currentlyHighlighted.element);

      wrapper.vm.scrollToLine(hash);

      await nextTick();
      expect(wrapper.vm.highlightedLine).toBe(linetoBeHighlighted.element);
      expect(currentlyHighlighted.classes()).not.toContain(HIGHLIGHT_CLASS_NAME);
      expect(linetoBeHighlighted.classes()).toContain(HIGHLIGHT_CLASS_NAME);
    });
  });
});
