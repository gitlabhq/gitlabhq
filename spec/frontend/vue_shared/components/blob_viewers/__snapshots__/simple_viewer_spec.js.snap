// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blob Simple Viewer component rendering matches the snapshot 1`] = `
<div>
  <div
    class="code code-syntax-highlight-theme file-content gl-flex js-syntax-highlight"
  >
    <div
      class="!gl-px-0 line-numbers"
    >
      <div
        class="diff-line-num gl-flex line-links"
      >
        <a
          class="!gl-shadow-none -gl-mx-2 file-line-blame gl-select-none"
          href="/blame/foo-bar#L1"
        />
        <a
          class="!gl-shadow-none file-line-num gl-select-none"
          data-line-number="1"
          href="#L1"
          id="reference-0"
        >
          1
        </a>
      </div>
      <div
        class="diff-line-num gl-flex line-links"
      >
        <a
          class="!gl-shadow-none -gl-mx-2 file-line-blame gl-select-none"
          href="/blame/foo-bar#L2"
        />
        <a
          class="!gl-shadow-none file-line-num gl-select-none"
          data-line-number="2"
          href="#L2"
          id="reference-1"
        >
          2
        </a>
      </div>
      <div
        class="diff-line-num gl-flex line-links"
      >
        <a
          class="!gl-shadow-none -gl-mx-2 file-line-blame gl-select-none"
          href="/blame/foo-bar#L3"
        />
        <a
          class="!gl-shadow-none file-line-num gl-select-none"
          data-line-number="3"
          href="#L3"
          id="reference-2"
        >
          3
        </a>
      </div>
    </div>
    <div
      class="blob-content gl-flex gl-flex-col gl-overflow-y-auto gl-w-full"
    >
      <pre
        class="!gl-p-0 code highlight"
      >
        <code
          data-blob-hash="foo-bar"
        >
          <span
            id="reference-3"
          >
            First
          </span>
          <span
            id="reference-4"
          >
            Second
          </span>
          <span
            id="reference-5"
          >
            Third
          </span>
        </code>
      </pre>
    </div>
  </div>
</div>
`;
